### PR TITLE
[WIP] Optimize helper functions

### PR DESCRIPTION
**Pull request by nathanmlim**
Some minor code optimizations to some functions, particularly when dealing with unit.Quantity objects 

Still working through some of the functions, moving them around, and testing before committing.

### CODE REVIEW

**Comment by sgill2** [comment]
This is good for you to do since it gives me motivation to put together some much needed unit tests. Let me put one together first (before I accept this change and your future ones) before I approve the change.

**Comment by davidlmobley** [comment]
@sgill2 - unit tests are indeed much needed. You should be able to bring some in via a separate PR and get them checked/merged before this is merged; it will be especially important to have unit tests if code reorganization is going on, since it's easy for that to break things.

**Comment by davidlmobley** [comment]
@limn1 - if you have suggestions for units tests (e.g. tests you're running before committing) let us know.  Travis-CI is running on this repo and will basically run any tests in the `blues/tests` directory (specifically, functions and/or classes within modules there that begin with the word `test`) on any new pull requests and report back on whether the test functions raise any exceptions. 

Tests here are quite minimal, but it's possible to have extensive tests. For example, see https://github.com/open-forcefield-group/smarty/blob/master/smarty/tests/test_forcefield.py which tests a wide variety of things about the SMIRFF forcefields, ranging from whether the modules run properly and yield finite energies, to whether they yield the CORRECT energies for specific systems and handle problem cases correctly, etc. It's even possible to have tests that the intended exceptions are raised in specific problem cases.

The main ideas I try to follow are this:
a) Anytime you write new code, write a test to ensure it runs properly (and also, when coding it, make sure it raises exceptions if anything goes wrong so the tests will more effectively catch those!)
b) Anytime you run into a bug/problem/issue with old code that isn't caught by one of the existing tests, add a new test or extend/modify an old test so that it will catch that and related issues in the future
c) Anytime you can come up with a case where you know exactly what should happen, write a test that checks if indeed that happens

Basically, we want to get to the place where, if a pull request will break something, the testing infrastructure will tell us. :)

**Comment by limn1** [comment]
@davidlmobley I can't say that I have much experience using unit tests. I've really only played around with them in the Orion cubes. The tests I'm doing so far aren't fairly extensive, more so that I don't it doesn't raise any errors.

@sgill2 I think you can go ahead and close your PR #23, I've included your credit additions into my latest commit https://github.com/MobleyLab/blues/pull/24/commits/45f82a100142a3544d65a9ae85de02bdf289c7e2 done additional formatting. Also, by opening this PR I meant to just show the kind of changes we talked about on Friday. As mentioned, I'm intending on moving things around/renaming functions so that they fall more in like with the OpenMM scheme. I'll request for you to review once I've finished doing all the changes :smile:

File: blues/ncmc.py [hunk 1]
```diff
@@ -181,59 +189,45 @@ def __init__(self, temperature, residueList, **kwds):
...
         """
         for index in (atomList):
             system.setParticleMass(index, 0*unit.daltons)
-        
+
 
     def calculate_com(self,  pos_state, total_mass=None, mass_list=None, residueList=None, rotate=False):
         """
-        This function calculates the com of specified residues and optionally 
+        This function calculates the com of specified residues and optionally
```
Comment (davidlmobley) [inline blues/ncmc.py]: @sgill2 (and @nathanmlim) - can we not use "com" as short for "center of mass" without defining that the first time it's used in the doc string?

File: blues/ncmc.py [hunk 2]
```diff
@@ -345,15 +339,15 @@ def create_normalSystem(self, coord_file, top_file):
 #        self.dummy_integrator = openmm.openmm.LangevinIntegrator(temperature, friction, timestep)
 #        self.md_simulation = openmm.app.simulation.Simulation(topology=self.normalsystem.topology, system=self.normalsystem.system, integrator=self.md_integrator)
 #        self.dummy_simulation = openmm.app.simulation.Simulation(topology=self.normalsystem.topology, system=self.normalsystem.system, integrator=self.dummy_integrator)
-
```
Comment (davidlmobley) [inline blues/ncmc.py]: @sgill2 - if these spaces are coming from you (or from you, @nathanmlim ), can you set your editor to strip trailing spaces? (Harass me on Slack if you don't know how). It's best to all use the same settings for editing text files so GitHub doesn't show changes when in fact there are no changes.

File: blues/posedart.py [hunk 3]
```diff
@@ -72,8 +80,8 @@ def setDartUpdates(self, residueList):
         self.residueList = residueList
 
     def defineLigandAtomsFromFile(lig_resname, coord_file, top_file=None):
```
Comment (davidlmobley) [inline blues/posedart.py]: This function seems to need a doc string.

File: blues/posedart.py [hunk 4]
```diff
@@ -276,12 +284,12 @@ def poseMove(self, context=None, residueList=None):
...
+        """
         For dynamically updating dart positions based on positions
         of other particles.
         This takes the weighted average of the specified particles
         and changes the dartboard of the object
-        
+
         Arguments
         ---------
         particle_pairs: list of list of ints
```
Comment (davidlmobley) [inline blues/posedart.py]: Ints of what? Atom indices? Doc strings should have enough info that I could actually use the function without having to read the rest of the code.

**Comment by davidlmobley** [comment]
Guys - I was just scrolling some of the code and noticed I'm not at all happy with the doc strings. I realize that documentation tends to be the last thing we're inclined to do, but I really want us to shift towards a model where we write the doc string first. I've managed to do this for much of my coding these days and it actually helps me a lot. I spend a while thinking about what arguments my code will take in (with what data types) and return, then write the doc string, and I sometimes spend more time doing that then I do writing the actual code. It actually makes writing the code much easier.

You want to strive towards doc strings that are complete enough that I could use your function for something even if separated from the rest of the code in the package (e.g. if I want to reuse it for something else) or so that I can figure out how to do something with it without knowing anything about the rest of the package.

Usually this means that you should provide a complete list of input parameters and returned parameters (in separate sections) with, for each, what data type it is expected to be (list of ints, etc.) then a brief description of what is supposed to be contained there ("indices of atoms to be considered for ...."). Depending on how involved the function is, you might also need a brief description of what it does before you provide the list of arguments/returns, and some usage notes which come after the list of arguments/returns. 

Make sure also that if you use abbreviations ("com" for "center of mass") those are defined.

File: blues/ncmc.py (hunk 2 above)
Comment (nathanmlim) [inline blues/ncmc.py]: These line strips are probably coming from my editor.

**Comment by nathanmlim** [comment]
@davidlmobley I do agree that our docstrings are currently lacking but I find that during development the docstrings get in the way and take up precious screen space that I opt to leave them for last. I'm still in the middle of refactoring the core code (see the [ipython](https://github.com/MobleyLab/blues/blob/optimizations/notebooks/ClassRefactoring.ipynb) notebooks) so please do not take what is currently on this branch as a reflection of the final PR.

**Comment by davidlmobley** [comment]
@nathanmlim  - OK, if it's easier for you to work that way it's totally fine. :) 

Sometimes it can be helpful to add a checklist to your PR so I know that you're planning to do that later (since the box is unchecked), in which case I'll just leave you alone until you check the box. :) 

Checklist format is like this:
```
- [x] Refactor ABC to ...
- [ ] Refactor XYZ to ...
- [ ] Revise doc strings to reflect refactorization of ...
```

If you put this in the first entry of your pull request (which, BTW, you can edit as you go to update the checklist/add more items) GitHub will show a handy summary of how many checkboxes are checked off in the PR when you look at the PR in the list of PRs. :)

**Comment by nathanmlim** [comment]
Closing this as I've put it all into another branch.
